File: elixir/phone-number/phone_number.exs
defmodule Phone do
  @length 10
  @bad_number String.duplicate("0", @length)

  def number(digits) do
    digits = scrub_formatting(digits)

    cond do
      String.length(digits) == @length ->
        digits
      String.length(digits) == @length + 1 && valid_country_code?(digits) ->
        remove_country_code(digits)
      true ->
        @bad_number
    end
  end

  def pretty(digits) do
    "(#{area_code(digits)}) #{exchange(digits)}-#{suffix(digits)}"
  end

  def area_code(digits), do: number(digits) |> String.slice(0, 3)

  defp exchange(digits), do: number(digits) |> String.slice(3, 3)

  defp suffix(digits), do: number(digits) |> String.slice(6, 4)

  defp valid_country_code?(digits), do: String.starts_with?(digits, "1")

  defp remove_country_code(digits), do: String.slice(digits, 1, 10)

  defp scrub_formatting(digits), do: Regex.replace(%r/\D+/, digits, "")
end

File: ruby/phone-number/phone_number.rb
class PhoneNumber
  LENGTH = 10
  BAD_NUMBER = '0' * LENGTH

  attr_reader :number

  def initialize(input)
    @number = extract_number(input)
  end

  def area_code
    @area_code ||= number.slice(0, 3)
  end

  def to_s
    "(#{area_code}) #{exchange}-#{suffix}"
  end

  private

  def extract_number(input)
    input = input.gsub(/\D/, '')

    if input.length == LENGTH
      input
    elsif input.length == LENGTH + 1 && valid_country_code?(input)
      remove_country_code(input)
    else
      BAD_NUMBER
    end
  end

  def valid_country_code?(input)
    input.start_with?('1')
  end

  def remove_country_code(input)
    input.slice(1..-1)
  end

  def exchange
    @exchange ||= number.slice(3, 3)
  end

  def suffix
    @suffix ||= number.slice(6, 4)
  end
end
